libbpf-cargo/src/lib.rs

```diff
@@ -46,5 +55,164 @@
 //! build`. This is a convenience command so you don't forget any steps. Alternatively, you could
 //! write a Makefile for your project.
 
-#[doc(hidden)]
-pub fn foo() {}
+use std::path::{Path, PathBuf};
+use std::result;
+
+use tempfile::tempdir;
+use thiserror::Error;
+
+// libbpf-cargo binary is the primary consumer of the following modules. As such,
+// we do not use all the symbols. Silence any unused code warnings.
+#[allow(dead_code)]
+mod btf;
+#[allow(dead_code)]
+mod build;
+#[allow(dead_code)]
+mod gen;
+#[allow(dead_code)]
+mod make;
+#[allow(dead_code)]
+mod metadata;
+
+#[cfg(test)]
+mod test;
+
+/// Canonical error type for this crate.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Error building BPF object file: {0}")]
+    Build(String),
+    #[error("Error generating skeleton: {0}")]
+    Generate(String),
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// `SkeletonBuilder` builds and generates a single skeleton.
+///
+/// This interface is meant to be used in build scripts.
+///
+/// # Examples
+///
+/// ```no_run
+/// use libbpf_cargo::SkeletonBuilder;
+///
+/// SkeletonBuilder::new("myobject.bpf.c")
+///     .debug(true)
+///     .clang("/opt/clang/clang")
+///     .generate("/output/path")
+///     .unwrap();
+/// ```
+pub struct SkeletonBuilder {
+    debug: bool,
+    source: PathBuf,
+    clang: PathBuf,
+    clang_options: String,
+    skip_clang_version_check: bool,
+    rustfmt: PathBuf,
+}
+
+impl SkeletonBuilder {
+    /// Create a new builder instance, where `source` is the path to the BPF object source
+    /// (typically suffixed by `.bpf.c`)
+    pub fn new<P: AsRef<Path>>(source: P) -> Self {
+        SkeletonBuilder {
+            debug: false,
+            source: source.as_ref().to_path_buf(),
+            clang: "clang".into(),
+            clang_options: String::new(),
+            skip_clang_version_check: false,
+            rustfmt: "rustfmt".into(),
+        }
+    }
+
+    /// Turn debug output on or off
+    ///
+    /// Default is off
+    pub fn debug(&mut self, debug: bool) -> &mut SkeletonBuilder {
+        self.debug = debug;
+        self
+    }
+
+    /// Specify which `clang` binary to use
+    ///
+    /// Default searchs `$PATH` for `clang`
+    pub fn clang<P: AsRef<Path>>(&mut self, clang: P) -> &mut SkeletonBuilder {
+        self.clang = clang.as_ref().to_path_buf();
+        self
+    }
+
+    /// Pass additional options to `clang` when buildling BPF object file
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use libbpf_cargo::SkeletonBuilder;
+    ///
+    /// SkeletonBuilder::new("myobject.bpf.c")
+    ///     .options("-DMACRO=value -I/some/include/dir")
+    ///     .generate("/output/path")
+    ///     .unwrap();
+    /// ```
+    pub fn options<S: AsRef<str>>(&mut self, opts: S) -> &mut SkeletonBuilder {
```

Comment:
am I only one confused whose options it is? {compiler,clang}_{options,args,extra_args} seems more self-describing?

Reply:
Maybe `clang_args` would make the most sense